seaborn/matrix.py

```diff
@@ -708,9 +715,16 @@ def __init__(self, data, pivot_kws=None, z_score=None, standard_scale=None,
         self.fig = plt.figure(figsize=figsize)
 
         if row_colors is not None:
+            if isinstance(row_colors, (pd.DataFrame, pd.Series)):
+                # Ensure colors match data indices
+                row_colors = row_colors.ix[data.index]
```

Comment:
I think this is reasonable, but you'll want to be very careful to explicitly test the expected failure modes.